### PR TITLE
Support On-Demand Resource 

### PR DESCRIPTION
**Pull request by JunyiXie**
https://github.com/CocoaPods/CocoaPods/issues/9606

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* Support On-Demand Resource   
  [JunyiXie](https://github.com/JunyiXie)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/xcodeproj/project/object/native_target.rb [hunk 1]
```diff
@@ -552,6 +552,36 @@ def add_resources(resource_file_references)
           end
         end
 
+        # remove on demand resource files to the resources build phase of the target.
```
Comment (dnkoutso) [inline lib/xcodeproj/project/object/native_target.rb]: nit capital case 'R'

**Comment by dnkoutso** [comment]
Thanks, can you please add tests and a changelog entry?

**Comment by dnkoutso** [comment]
Please also run `bundle exec rubocop -a` to fix Rubocop issues for this to go green.

File: lib/xcodeproj/project/object/native_target.rb [hunk 2]
```diff
@@ -552,6 +552,36 @@ def add_resources(resource_file_references)
...
+        #
+        # @param  [Array<PBXFileReference>] on_demand_resource_file_references
+        #         the files references of the on demand resources to the target.
+        #
+        # @return [void]
+        #
+        def remove_on_demand_resources(on_demand_resource_file_references)
+          on_demand_resource_file_references.each do |file|
```
Comment (dnkoutso) [inline lib/xcodeproj/project/object/native_target.rb]: I guess this is already defined? `on_demand_resource_file_references`?

Comment (dnkoutso) [reply lib/xcodeproj/project/object/native_target.rb]: nm I am stupid and this is a parameter. Hmmmm this doesn't really match what the function does. Its just removing resource file reference which can be done in `remove_file_reference` inside `build_phase.rb`. I don't think this method is really needed.

Comment (dnkoutso) [reply lib/xcodeproj/project/object/native_target.rb]: Or at least can we assert that `ASSET_TAGS` is present on these resources?

**Comment by shirakaba** [comment]
To provide more context: @JunyiXie has written an excellent [blog post](https://junyixie.github.io/2020/03/19/cocoapods-new-feature/) summarising the changes necessary to `CocoaPods/Core` (https://github.com/CocoaPods/Core/pull/622) and `CocoaPods/Xcodeproj` (this PR) in order to support On-Demand Resources. It's quite understandable [with Google Translate](https://translate.google.co.uk/translate?sl=zh-CN&tl=en&u=https%3A%2F%2Fjunyixie.github.io%2F2020%2F03%2F19%2Fcocoapods-new-feature%2F).

@JunyiXie I notice that this PR is waiting on further changes. If any help is needed with communication with @dnkoutso , I'd be happy to (try to) translate.

@JunyiXie 我注意到这个PR等一步改变。如果有任何沟通问题跟@dnkoutso，我愿意试试帮助翻译。

File: lib/xcodeproj/project/object/native_target.rb [hunk 3]
```diff
@@ -552,6 +552,36 @@ def add_resources(resource_file_references)
...
+        #         the files references of the on demand resources to the target.
+        #
+        # @return [void]
+        #
+        def add_on_demand_resources(on_demand_resource_file_references)
+          on_demand_resource_file_references.each do |file|
+            next if resources_build_phase.include?(file)
+            build_file = project.new(PBXBuildFile)
+            build_file.file_ref = file
+            build_file.settings = (build_file.settings ||= {}).merge({"ASSET_TAGS" => ["OnDemand"]})
```
Comment (amorde) [inline lib/xcodeproj/project/object/native_target.rb:580]: I don't think we want to be hard-coding asset tags here. According to [Apple's docs](https://developer.apple.com/library/archive/documentation/FileManagement/Conceptual/On_Demand_Resources_Guide/Tagging.html#//apple_ref/doc/uid/TP40015083-CH3-SW1) tags are supposed to be manually specified, and those tags are then used in code to [download the assets](https://developer.apple.com/documentation/foundation/nsbundleresourcerequest)

**Comment by JunyiXie** [comment]
I'm sorry I didn't read the GitHub news before, I will improve it in the next few days.

File: lib/xcodeproj/project/object/native_target.rb (hunk 2 above)
Comment (JunyiXie) [reply lib/xcodeproj/project/object/native_target.rb]: emmm... Where do I write it more reasonable?

File: lib/xcodeproj/project/object/native_target.rb (hunk 3 above)
Comment (JunyiXie) [reply lib/xcodeproj/project/object/native_target.rb]: thanks, now use the podspec on_demand_resources specified tag 
https://github.com/CocoaPods/Xcodeproj/pull/742/commits/81403b6248a44e0ad15c151680cb22360d06141c#diff-5ee67a82fd5ae8aa882ddaf26476e138R577

**Comment by JunyiXie** [comment]
> To provide more context: @JunyiXie has written an excellent [blog post](https://junyixie.github.io/2020/03/19/cocoapods-new-feature/) summarising the changes necessary to `CocoaPods/Core` ([CocoaPods/Core#622](https://github.com/CocoaPods/Core/pull/622)) and `CocoaPods/Xcodeproj` (this PR) in order to support On-Demand Resources. It's quite understandable [with Google Translate](https://translate.google.co.uk/translate?sl=zh-CN&tl=en&u=https%3A%2F%2Fjunyixie.github.io%2F2020%2F03%2F19%2Fcocoapods-new-feature%2F).
> 
> @JunyiXie I notice that this PR is waiting on further changes. If any help is needed with communication with @dnkoutso , I'd be happy to (try to) translate.
> 
> @JunyiXie 我注意到这个PR等一步改变。如果有任何沟通问题跟@dnkoutso，我愿意试试帮助翻译。

**Comment by ignkarman** [comment]
Is there a plan to get this feature included in a future release? My team would love to make use of this functionality

**Comment by dnkoutso** [comment]
@JunyiXie i would like to move this feature forward, can you rebase to latest master and add tests? if not I can take over this PR and give you credit for it so I can land it.

**Comment by dnkoutso** [comment]
superseded by https://github.com/CocoaPods/Xcodeproj/pull/844 due to inactivity. giving credit to original author as well.

**Pull request closed without merging**